db/compaction/compaction_job.cc

```diff
@@ -353,35 +356,75 @@ void CompactionJob::GenSubcompactionBoundaries() {
   // Group the ranges into subcompactions
   const double min_file_fill_percent = 4.0 / 5;
   int base_level = v->storage_info()->base_level();
+  uint64_t max_subcompactions = static_cast<uint64_t>(c->max_subcompactions());
   uint64_t max_output_files = static_cast<uint64_t>(std::ceil(
       sum / min_file_fill_percent /
       MaxFileSizeForLevel(
           *(c->mutable_cf_options()), out_lvl,
           c->immutable_options()->compaction_style, base_level,
           c->immutable_options()->level_compaction_dynamic_level_bytes)));
-  uint64_t subcompactions =
-      std::min({static_cast<uint64_t>(ranges.size()),
-                static_cast<uint64_t>(c->max_subcompactions()),
-                max_output_files});
+  uint64_t subcompactions = std::min({static_cast<uint64_t>(ranges.size()),
+                                      max_subcompactions, max_output_files});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // The number of files in the start level cannot be larger than
+    // ranges.size() since ranges also contains the number of files from other
+    // levels Note that mutex has been locked, so it is safe to access
```

Comment:
I'd prefer explicitly asserting mutex being locked like `db_mutex_->AssertHeld();` in code instead of merely putting a note in the comment.

Reply:
will do.